File: Assets/Bundles/Server/Resources/TS/Services/Global/Shutdown/ShutdownService.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local ServerSignals = require("Server/TS/ServerSignals").ServerSignals
local RunUtil = require("Shared/TS/Util/RunUtil").RunUtil
local SetInterval = require("Shared/TS/Util/Timer").SetInterval
local ShutdownService
do
	ShutdownService = setmetatable({}, {
		__tostring = function()
			return "ShutdownService"
		end,
	})
	ShutdownService.__index = ShutdownService
	function ShutdownService.new(...)
		local self = setmetatable({}, ShutdownService)
		return self:constructor(...) or self
	end
	function ShutdownService:constructor(playerService)
		self.playerService = playerService
		self.playerConnected = false
		self.timeWithNoPlayers = 0
	end
	function ShutdownService:OnStart()
		ServerSignals.PlayerJoin:connect(function(event)
			self.playerConnected = true
			self.timeWithNoPlayers = 0
		end)
		local intervalTime = 10
		SetInterval(intervalTime, function()
			if #self.playerService:GetPlayers() == 0 then
				self.timeWithNoPlayers += intervalTime
				if RunUtil:IsEditor() then
					return nil
				end
				if self.playerConnected then
					if self.timeWithNoPlayers >= ShutdownService.SHUTDOWN_TIME_ALL_PLAYERS_LEFT then
						print("Server will shutdown due to excessive time with all players having left.")
						self:Shutdown()
					end
				else
					if self.timeWithNoPlayers >= ShutdownService.SHUTDOWN_TIME_NOBODY_CONNECTED then
						print("Server will shutdown due to excessive time with nobody ever connecting.")
						self:Shutdown()
					end
				end
			end
		end)
	end
	function ShutdownService:Shutdown()
		local serverBootstrap = GameObject:Find("ServerBootstrap"):GetComponent("ServerBootstrap")
		serverBootstrap:Shutdown()
	end
	ShutdownService.SHUTDOWN_TIME_NOBODY_CONNECTED = 10 * 60
	ShutdownService.SHUTDOWN_TIME_ALL_PLAYERS_LEFT = 1 * 60
end
-- (Flamework) ShutdownService metadata
Reflect.defineMetadata(ShutdownService, "identifier", "Bundles/Server/Services/Global/Shutdown/ShutdownService@ShutdownService")
Reflect.defineMetadata(ShutdownService, "flamework:parameters", { "Bundles/Server/Services/Global/Player/PlayerService@PlayerService" })
Reflect.defineMetadata(ShutdownService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(ShutdownService, "$:flamework@Service", Service, { {} })
return {
	ShutdownService = ShutdownService,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Server/Services/Global/Shutdown/ShutdownService.ts
import { OnStart, Service } from "@easy-games/flamework-core";
import { ServerSignals } from "Server/ServerSignals";
import { RunUtil } from "Shared/Util/RunUtil";
import { SetInterval } from "Shared/Util/Timer";
import { PlayerService } from "../Player/PlayerService";

@Service({})
export class ShutdownService implements OnStart {
	private playerConnected = false;
	private timeWithNoPlayers = 0;

	private static SHUTDOWN_TIME_NOBODY_CONNECTED = 10 * 60;
	private static SHUTDOWN_TIME_ALL_PLAYERS_LEFT = 1 * 60;

	constructor(private readonly playerService: PlayerService) {}

	OnStart(): void {
		ServerSignals.PlayerJoin.Connect((event) => {
			this.playerConnected = true;
			this.timeWithNoPlayers = 0;
		});

		const intervalTime = 10;
		SetInterval(intervalTime, () => {
			if (this.playerService.GetPlayers().size() === 0) {
				this.timeWithNoPlayers += intervalTime;

				if (RunUtil.IsEditor()) {
					return;
				}

				if (this.playerConnected) {
					if (this.timeWithNoPlayers >= ShutdownService.SHUTDOWN_TIME_ALL_PLAYERS_LEFT) {
						print("Server will shutdown due to excessive time with all players having left.");
						this.Shutdown();
					}
				} else {
					if (this.timeWithNoPlayers >= ShutdownService.SHUTDOWN_TIME_NOBODY_CONNECTED) {
						print("Server will shutdown due to excessive time with nobody ever connecting.");
						this.Shutdown();
					}
				}
			}
		});
	}

	public Shutdown(): void {
		const serverBootstrap = GameObject.Find("ServerBootstrap").GetComponent<ServerBootstrap>();
		serverBootstrap.Shutdown();
	}
}
